packages/grafana-runtime/src/utils/useFavoriteDatasources.ts

```diff
@@ -0,0 +1,120 @@
+import { useCallback, useEffect, useState } from 'react';
+
+import { DataSourceInstanceSettings } from '@grafana/data';
+
+import { UserStorage } from './userStorage';
+
+const FAVORITE_DATASOURCES_KEY = 'favoriteDatasources';
+
+/**
+ * A hook for managing favorite data sources using user storage.
+ * This hook provides functionality to store and retrieve a list of favorite data source UIDs
+ * using the backend user storage (with localStorage fallback).
+ *
+ * @returns An object containing:
+ * - An array of favorite data source UIDs
+ * - An array of favorite data source UIDs that were initially loaded from storage
+ * - A function to add a data source to favorites
+ * - A function to remove a data source from favorites
+ * - A function to check if a data source is favorited
+ * - A function to toggle a data source's favorite status
+ * @public
+ */
+export function useFavoriteDatasources(): {
+  favoriteDatasources: string[];
+  initialFavoriteDataSources: string[];
+  addFavoriteDatasource: (ds: DataSourceInstanceSettings) => void;
+  removeFavoriteDatasource: (ds: DataSourceInstanceSettings) => void;
+  isFavoriteDatasource: (dsUid: string) => boolean;
+  toggleFavoriteDatasource: (ds: DataSourceInstanceSettings) => void;
+} {
+  const [userStorage] = useState(() => new UserStorage('grafana-runtime'));
+  const [favoriteDatasources, setFavoriteDatasources] = useState<string[]>([]);
+  const [initialFavoriteDataSources, setInitialFavoriteDataSources] = useState<string[]>([]);
+
+  // Load favorites from storage on mount
+  useEffect(() => {
+    const loadFavorites = async () => {
+      try {
+        const stored = await userStorage.getItem(FAVORITE_DATASOURCES_KEY);
+        if (stored) {
+          const parsed = JSON.parse(stored);
+          if (Array.isArray(parsed)) {
+            setFavoriteDatasources(parsed);
+            setInitialFavoriteDataSources(parsed);
+          }
+        }
+      } catch (error) {
+        console.warn('Failed to load favorite data sources:', error);
+        setFavoriteDatasources([]);
+      }
+    };
+
+    loadFavorites();
+  }, [userStorage]);
+
+  // Helper function to save favorites to storage
+  const saveFavorites = useCallback(
+    async (newFavorites: string[]) => {
+      try {
+        await userStorage.setItem(FAVORITE_DATASOURCES_KEY, JSON.stringify(newFavorites));
+        setFavoriteDatasources(newFavorites);
+      } catch (error) {
+        console.warn('Failed to save favorite data sources:', error);
+      }
+    },
+    [userStorage]
+  );
```

Comment:
Its kind of beside the point for this task, but I find all this boilerplate for what I would guess would be pretty common usage of UserStorage to be pretty cumbersome. As a consumer of it, it would expect to have all of this handled for me.

Something like:

```ts
const [favorites, setFavorites] = useUserStorage("grafana-runtime/favorite-datasources");

const addFavorite = useCallback((newFav) => {
  setFavorites(v => [...v, newFav])
}, [setFavorites])

// etc

return {
  favorites,
  addFavorite 
}

Reply:
actually - it might be nice to instead generate the RKT Query API client for this and have a wrapper around those query/mutation hooks? We get a load of lifecycle hooks for free if we do that!